pkg/csi/service/wcp/controller.go

```diff
@@ -1021,66 +1021,74 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
-							"failed to get the PodVM Moref from the PodVM UUID: %s in datacenter: %s with err: %+v",
-							v, dcMorefValue, err)
-					}
-					isStillAttached := false
-					timeout := 4 * time.Minute
-					pollTime := time.Duration(5) * time.Second
-					err = wait.Poll(pollTime, timeout, func() (bool, error) {
-						diskUUID, err := cnsvolume.IsDiskAttached(ctx, podVM, req.VolumeId, true)
+						var vCenterHost, dcMorefValue string
+						for key, value := range vcdcMap {
+							vCenterHost = key
+							dcMorefValue = value
+						}
+						vc, err := c.manager.VcenterManager.GetVirtualCenter(ctx, vCenterHost)
+						if err != nil {
+							return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+								"cannot get virtual center %s from virtualcentermanager while attaching disk with error %+v",
+								vc.Config.Host, err)
+						}
+						// Connect to VC.
+						err = vc.Connect(ctx)
 						if err != nil {
-							log.Infof("retrying the IsDiskAttached check again for volumeId %q. Err: %+v", req.VolumeId, err)
-							return false, nil
+							return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+								"failed to connect to Virtual Center: %s", vc.Config.Host)
 						}
-						if diskUUID != "" {
-							log.Infof("diskUUID: %q is still attached. Retrying in 5 seconds.", diskUUID)
-							isStillAttached = true
-							return false, nil
+						podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
+						if err != nil {
+							if err == cnsvsphere.ErrVMNotFound {
+								log.Infof("virtual machine not found for vmUUID %q. "+
+									"Thus, assuming the volume is detached.", v)
+								break
+							}
+							return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+								"failed to get the PodVM Moref from the PodVM UUID: %s in datacenter: %s with err: %+v",
+								v, dcMorefValue, err)
 						}
-						return true, nil
-					})
-					if err != nil {
-						if isStillAttached {
-							log.Errorf("volume %q is still attached to node %q", req.VolumeId, req.NodeId)
+						isStillAttached := false
+						timeout := 4 * time.Minute
+						pollTime := time.Duration(5) * time.Second
+						err = wait.Poll(pollTime, timeout, func() (bool, error) {
+							diskUUID, err := cnsvolume.IsDiskAttached(ctx, podVM, req.VolumeId, true)
+							if err != nil {
+								log.Infof("retrying the IsDiskAttached check again for volumeId %q. Err: %+v", req.VolumeId, err)
+								return false, nil
+							}
+							if diskUUID != "" {
+								log.Infof("diskUUID: %q is still attached to podVM %+v. Retrying in 5 seconds.", podVM, diskUUID)
+								isStillAttached = true
+								return false, nil
+							}
+							return true, nil
+						})
+						if err != nil {
+							if isStillAttached {
+								log.Errorf("volume %q is still attached to node %q and podVM %+v", req.VolumeId, req.NodeId, podVM)
```

Comment:
Will this print the whole podVM object? Just thinking that the error log doesn't become verbose, and if just printing vm name and vm Moid should suffice.